### ISSUE TITLE
[Bug]: Compression Policy Stalling After Upgrade to TimescaleDB 2.16.2

### ISSUE THREAD

**Issue by rdmolony** [timescale/timescaledb#7502]
### What type of bug is this?

Crash

### What subsystems and features are affected?

Compression

### What happened?

**TL;DR;** -

My database's compression policy job (`CALL _timescaledb_functions.policy_compression()`) has not run successfully since I upgraded from version 2.15.3 to 2.16.2, approximately four months ago. It appears to get stuck on specific chunks during compression. I seem to have fixed specific chunks by manually decompressing/compressing then, and I'm now running a query to do so for all chunks.  See the SQL query at the end of the longer description for more info. 

---

Recently, I noticed that the periodic job `CALL _timescaledb_functions.policy_compression()` was running indefinitely.

So I looked at this job's stats via ...

```sql
select * from timescaledb_information.job_stats
```

|job_id|application_name|schedule_interval|max_runtime|max_retries|retry_period|proc_schema|proc_name|owner|scheduled|fixed_schedule|config|next_start|initial_start|hypertable_schema|hypertable_name|check_schema|check_name|
|------|----------------|-----------------|-----------|-----------|------------|-----------|---------|-----|---------|--------------|------|----------|-------------|-----------------|---------------|------------|----------|
|1014|Compression Policy [1014]|3 days 12:00:00|00:00:00|-1|01:00:00|_timescaledb_functions|policy_compression|postgres|true|false|{"hypertable_id": 61, "compress_after": "7 days"}|-infinity||public|sensor_reading_generic|_timescaledb_functions|policy_compression_check| 

... and saw that the last time it  ran successfully was 4 months ago.

This timing corresponds with an upgrade of Timescale from version 2.15.3 to 2.16.2

I noticed that the compression policy query `CALL _timescaledb_functions.policy_compression()` was getting stuck on one particular hypertable chunk by running ...

```sql
with query as (
  select pid
    from pg_stat_activity
   where query = 'CALL _timescaledb_functions.policy_compression()'
)
select
    l.relation::regclass AS locked_relation,
    (chunk_schema || '.' || chunk_name)::regclass,
    c.range_start,
    l.mode,
    l.locktype,
    a.query,
    a.state,
    a.wait_event_type,
    a.wait_event
from
    pg_stat_activity a
    join pg_locks l  on l.pid = a.pid
    join query q on q.pid = l.pid
    left join timescaledb_information.chunks c on (chunk_schema || '.' || chunk_name)::regclass = l.relation::regclass
```

... while it was executing.

> **Note** - the compression policy holds a lock on each hypertable it is compressing.  So long as it is stuck on a specific hypertable, backfilling to this hypertable (i.e. importing historic readings) doesn't work work either, since INSERTs are blocked by this lock.
> 
> This hasn't been a major issue since backfilled historic data is mostly non-critical.

I found a relevant similar issue (thanks to `ChatGPT 4o`) that suggested manually compressing the chunk ...

> [Timescale Community Forum](https://www.timescale.com/forum/t/all-compression-policy-jobs-fail/756): In a forum post, a user described that all their compression policy jobs had been failing for approximately six weeks. The failures started after upgrading the Timescale extension. Manually running the jobs succeeded, but scheduled executions failed. The user sought advice on obtaining more detailed error information to diagnose the problem. ￼

So I tried this via ...

```sql
select compress_chunk('_timescaledb_internal.<chunk_name>')
```

... and found that the query ran indefinitely.

So (dumbly) I tried decompressing the chunk manually via ...

```sql
select decompress_chunk('_timescaledb_internal.<chunk_name>')
```

... and **it eventually decompressed after 40 minutes!**  Then I recompressed it, which this time took less than a minute!

**I don't understand quite what went wrong or how this fixed it.  `ChatGPT 4o` believed that this decompress/compress resolved metadata inconsistencies & bloat** - but my understanding is not good enough to dig into that.

So I re-run the compression policy, this time it got stuck on another chunk.

So now I'm running decompress/compress on all chunks via ...

```sql
DO $$
DECLARE
    chunk RECORD;
BEGIN
    FOR chunk IN
        SELECT chunk_schema, chunk_name
        FROM timescaledb_information.chunks
        WHERE is_compressed = true
    LOOP
        -- Step 1: Decompress the chunk
        RAISE NOTICE 'Decompressing chunk: %.%', chunk.chunk_schema, chunk.chunk_name;
        EXECUTE format('SELECT decompress_chunk(''%I.%I'')', chunk.chunk_schema, chunk.chunk_name);

        -- Step 2: Compress the chunk
        RAISE NOTICE 'Compressing chunk: %.%', chunk.chunk_schema, chunk.chunk_name;
        EXECUTE format('SELECT compress_chunk(''%I.%I'')', chunk.chunk_schema, chunk.chunk_name);
    END LOOP;
END $$;
```

Fingers crossed this resolves it

### TimescaleDB version affected

2.16.2

### PostgreSQL version used

PostgreSQL 14.13, compiled by Visual C++ build 1940, 64-bit

### What operating system did you use?

Windows Server 2019 Datacenter 17763.6532

### What installation method did you use?

Other

### What platform did you run on?

On prem/Self-hosted

### How can we reproduce the bug?

```bash
This might be hard to do other than upgrading from version 2.15.3 to 2.16.2 on a Windows server (if I've understood the issue)
```

**Comment by rdmolony** [comment]
I can confirm that after running the decompress/compress SQL query above compression jobs are now back running without any issues.

**Comment by erimatnor** [comment]
@rdmolony Thank you for reporting this issue. Do you think this is issue is specific to Windows?

**Comment by rdmolony** [comment]
Thanks @erimatnor, it could well be, I'm can't be sure

**Comment by normalnormie** [comment]
I was also stuck with "CALL _timescaledb_functions.policy_compression()"
I installed TimescaleDB some days ago and was not having issues until today
TimescaleDB version affected:
timescaledb-2-postgresql-14 (2.17.2~ubuntu22.04)
PostgreSQL version used:
(PostgreSQL) 14.12 (Ubuntu 14.12-0ubuntu0.22.04.1)
How can we reproduce the bug?
Followed https://docs.timescale.com/self-hosted/latest/install/installation-linux/ and inserted data for some days, the specific chunk that freezes policy_compression is from the biggest table and contains 57 million rows, also I had some out-of-space on disk past days that were solved before policy_compression started.

Used @rdmolony script but got stuck on: "SELECT decompress_chunk('_timescaledb_internal._hyper_25_27_chunk')" with a "Lock: relation"
Followed https://docs.timescale.com/use-timescale/latest/compression/compression-policy/ and paused the compression policy first by finding its job_id and then disabling it, script operation finished then enabled it back.

**Comment by normalnormie** [comment]
> I was also stuck with "CALL _timescaledb_functions.policy_compression()" I installed TimescaleDB some days ago and was not having issues until today TimescaleDB version affected: timescaledb-2-postgresql-14 (2.17.2~ubuntu22.04) PostgreSQL version used: (PostgreSQL) 14.12 (Ubuntu 14.12-0ubuntu0.22.04.1) How can we reproduce the bug? Followed https://docs.timescale.com/self-hosted/latest/install/installation-linux/ and inserted data for some days, the specific chunk that freezes policy_compression is from the biggest table and contains 57 million rows, also I had some out-of-space on disk past days that were solved before policy_compression started.
> 
> Used @rdmolony script but got stuck on: "SELECT decompress_chunk('_timescaledb_internal._hyper_25_27_chunk')" with a "Lock: relation" Followed https://docs.timescale.com/use-timescale/latest/compression/compression-policy/ and paused the compression policy first by finding its job_id and then disabling it, script operation finished then enabled it back.

An update: I detected the same problem but this time decided to wait some days to see if it solved on its own and it did, my app was stopped for a couple hours so it was probably new inserts blocking the compression job.

**Comment by antekresic** [comment]
Unfortunately, there are instances where compression policy can take a significant amount of time to re-compress a chunk which is what I think people are hitting here. This one of the things we are actively working on improving. One of the improvements is landing in January: https://github.com/timescale/timescaledb/pull/7482

Meanwhile, decompressing and compressing these problematic chunks manually is your best bet.

There are plans to work on lock contention here as well which can be caused by the compression policy and inserts into the chunk which should be compressed. But in the meantime, I suggest adjusting the policy to run on chunks which don't have data inserted into them anymore. For instance, if you are inserting data over the last 3 days, compression policy should compress chunks which are at least older than that. I realize this cannot be done for all workloads and that's why we plan on addressing this in the near future.

Hope this makes sense.

**Comment by antekresic** [comment]
Do you have any results testing version 2.18? It should improve the situation.

How do you know these chunks are corrupt? What leads you to believe that? Are you able to query the data in them after the compression policy fails?

Its hard to say how to fix this corruption without some more details.

As for monitoring compression policy, you can add verbose logging to the policy
```
SELECT alter_job(1000, config => '{"hypertable_id": 1, "compress_after": "7 days", "verbose_log": true}');
```
Reference:
https://docs.timescale.com/use-timescale/latest/user-defined-actions/alter-and-delete/#change-a-jobs-config

**Comment by rdmolony** [comment]
Hi @antekresic, I unfortunately cannot assist in this issue any longer as I no longer have access to the problematic database, and moreover once the issue was resolved the database was no longer in a problematic state.

(I'd love to have been able to create a reproducible issue, however, this issue depends upon database state which I am/was unable to share, good luck!)

**Comment by Dev10000** [comment]
Hi @antekresic 

> Do you have any results testing version 2.18? It should improve the situation.

I didn't manage to do much testing with version 2.18, but with 2.19, I've already noticed improvements. For example:
- I'm now able to multi-thread backfills on compressed chunks using timescaledb-parallel-copy without deadlocks
- I can also run multi-thread backfills on compressed chunks while compression is running, which previously caused issues
- These tests used data sorted by time in ascending order, starting from the oldest timestamps.

That said, my testing with 2.19 is still limited. I'm planning to test scenarios with completely unordered (mixed) input data, as well as concurrent parallel-copy runs from multiple sources, for example, 5 separate servers, each running parallel-copy with 6 workers, totaling 30 workers writing to the same Timescale table. All of these use cases have previously caused various issues, mainly deadlocks.

> How do you know these chunks are corrupt? What leads you to believe that? Are you able to query the data in them after the compression policy fails?

I believe a chunk is corrupt when the automatic compression policy hangs on it and won’t proceed. The only workaround so far has been to manually decompress and recompress the chunk, but sometimes even that doesn’t help, as in this case with version 2.17.0.

As mentioned in my earlier report, we're seeing errors like:

- type with OID 0 does not exist
- no index on 'insert_time' found for column range on chunk

In this specific case, the chunk is already uncompressed, but attempting to compress it fails with the OID error, even manual compression doesn’t work. So I believe the OID error might be what's causing the compression policy to freeze.

I'm able to query the data normally, but the manual compression fails due to the OID error. This makes me think something is broken in the chunk's metadata or catalog references..
 
> As for monitoring compression policy, you can add verbose logging to the policy
> ```
> SELECT alter_job(1000, config => '{"hypertable_id": 1, "compress_after": "7 days", "verbose_log": true}');
> ```

Thank you for this. I’ll try it out next time we encounter issues with the policy. So far, in our another test DB, it’s been working well with 2.19.

Just to mention: I’m doing this testing as time allows, in parallel with other unrelated development work, so I may not be able to respond immediately or test everything right away. But I’ll continue reporting anything notable as we go.

Thanks again for the support!
